README.md
# BregmaNet : Bregman Neural Networks

![license](https://img.shields.io/github/license/JordanFrecon/bregmanet)
![release](https://img.shields.io/github/v/release/JordanFrecon/bregmanet?include_prereleases)
![PyPI](https://img.shields.io/pypi/v/bregmanet)

[BregmaNet](https://github.com/JordanFrecon/bregmanet) is a PyTorch library providing multiple [Bregman Neural Networks](https://jordan-frecon.com/download/2022_Frecon_J_p-icml_bnn.pdf).
To date, implemented models cover Bregman variants of multi-layer perceptrons and various residual networks.


**Contributor:** Jordan Frécon (INSA Rouen Normandy, France)

## Table of Contents

1. [Requirements and Installation](#Requirements-and-Installation)
2. [Getting Started](#Getting-Started)
3. [List of Supported Models](#List-of-Supported-Models)
4. [Citation](#Citation)
5. [Contribution and Acknowledgments](#Contribution-and-Acknowledgments)



## Requirements and Installation

### :clipboard: Requirements

- PyTorch version >=1.7.1
- Python version >=3.6
- Torchvision version >=0.8.2


### :hammer: Installation

```
pip install bregmanet
```

In development versions can be found [here](https://test.pypi.org/project/bregmanet/).

## Getting Started

###  :warning: Precautions

* All images should be scaled within the domain range of the activation function.
* MLP models provided work only for 1-dimensional data inputs.
* MLP models are designed without a softmax final layer.
* All models need to be trained first. If you wish to provide your pretrained models, please [contribute](#Contribution-and-Acknowledgments).

### :rocket: Demos

Multiple demo files can be found [there](https://github.com/JordanFrecon/bregmanet) in the *demos* folder. It contains:
- *demo_toy_mlp.py*: training of MLP on the Two-spiral toy dataset.
- *demo_mnist_mlp.py*: training of MLP on the MNIST dataset.
- *demo_cifar10_resnet.py*: training of ResNet20 on the CIFAR-10 dataset.
- *demo_cifar100_resnet.py*: training of ResNet20 on the CIFAR-100 dataset.
- *demo_imagenet_resnet.py*: training of ResNet18 on the ImageNet dataset.



###  :page_with_curl: Loading a Model

To date, all Bregman neural models provided are not trained.
If needed, a training procedure is made available [there](https://github.com/JordanFrecon/bregmanet/) in the *demos/utils* folder.
In order to load a model, proceed as follows.

<details><summary>Multi-Layer Perceptrons</summary><p>

For a *sigmoid*-based MLP with 
- a linear input accepting 1d tensors of size 1024
- 3 hidden layers of size (1024, 1024, 512)
- a linear output layer mapping to 1d tensors of size 10

```python
import bregmanet
model = bregmanet.MLP(activation='sigmoid', input_dim=1024, hidden_dim=[1024, 1024, 512], output_dim=10)
```
</p></details>

<details><summary>ResNet</summary><p>

For a BregmanResNet20 with SoftPlus activation function:

```python
import bregmanet
model = bregmanet.bresnet20(activation='softplus')
```

</p></details>


## List of Supported Models

The following list reports all models currently supporting a Bregman variant. 
If you have any issue with one of them or wish to provide your own, please [contact us](mailto:jordan.frecon@gmail.com).

- MLP
- ResNet18
- ResNet20
- ResNet32
- ResNet34
- ResNet44
- ResNet56
- ResNet101
- Resnet110
- ResNet152
- Resnet1202
- ResNeXt50_32x4d
- ResNeXt101_32x8d
- WideResNet50_2
- WideResnet101_2


## Citation

If you use this package, please cite the following work:

```
@inproceedings{2022_Frecon_J_p-icml_bregmanet,
  title = {{Bregman Neural Networks}},
  author = {Frecon, Jordan and Gasso, Gilles and Pontil, Massimiliano and Salzo, Saverio},
  url = {https://hal.archives-ouvertes.fr/hal-03132512},
  series    = {Proceedings of Machine Learning Research},
  booktitle = {Proceedings of the 39th International Conference on Machine Learning,
               {ICML} 2022, 17-23 July 2022, Baltimore, USA},
  year = {2022},
}

```


## Contribution and Acknowledgments

Jordan Frecon would like to express his gratitude to the [Department of Computational Statistics and Machine Learning](https://www.iit.it/web/computational-statistics-and-machine-learning) (IIT, Genova, Italy) where part of this work was conducted during his postdoctoral position. The authors gratefully acknowledge the financial support of the French Agence Nationale de la Recherche (ANR), under grant ANR-20-CHIA-0021-01 ([project RAIMO](https://chaire-raimo.github.io)).

The proposed BregmanResNets for CIFAR-10 are based on a rework of the ResNet implementation of [Yerlan Idelbayev](https://github.com/akamaster/pytorch_resnet_cifar10).
Other ResNet models are devised by hinging upon the official PyTorch/TorchVision repository. For more information, please refer to:
- ResNet: ["Deep Residual Learning for Image Recognition"](https://arxiv.org/pdf/1512.03385.pdf) 
- ResNeXt: ["Aggregated Residual Transformation for Deep Neural Networks"](https://arxiv.org/pdf/1611.05431.pdf)
- WideResNet: ["Wide Residual Networks"](https://arxiv.org/pdf/1605.07146.pdf)

All kind of contributions are welcome, do not hesitate to [contact us!](mailto:jordan.frecon@gmail.com)


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## [v1.0.0](https://github.com/JordanFrecon/bregmanet/releases/tag/v1.0.0) (14/06/2022)

- First release of `bregmanet`!


## [v0.1.0](https://github.com/JordanFrecon/bregmanet/releases/tag/v0.1.0) (14/06/2022)

- First pre-release of `bregmanet`!


bregmanet/mlp.py
import torch
import torch.nn as nn
import bregmanet.utils.weight as weight
import bregmanet.utils.activation as act


class MLP(nn.Module):
    r"""
        Bregman Multilayer Perceptron

        Arguments:
            - activation_name (string): activation function.
            - input_dim (integer): size of the input data flattened into a 1d tensor.
            - hidden_dim (list of integer): number of hidden neurons per layer.
            - output_dim (integer): number of output neurons.
            - version (string): switch between 'standard' and 'bregman' MLP. (default: 'bregman')
            - init (string): switch between 'random' and 'deterministic' initialization of the weights and biases.
             (default: 'random')
            - weight_norm (boolean): if True, perform layer-wise weights normalization. (default: False)

        Example:
            model = MLP(activation_name='sigmoid', input_dim=1024, hidden_dim=[1024, 512, 1024], output_dim=10)

        """

    def __init__(self, activation, version='bregman', hidden_dim=None, input_dim=None, output_dim=1,
                 init='random', weight_norm=False):
        super().__init__()
        version = version.lower()
        activation = activation.lower()
        init_param = weight.parameter_initialization(version=version, init_type=init)
        if version == 'bregman':
            self.__class__.__name__ = 'BregmanMLP'

        # Parameters
        self.num_neurons = [input_dim] if hidden_dim is None else hidden_dim
        self.num_layers = self.num_neurons.__len__()
        self.weight_norm = weight_norm
        self.version = version

        # Hidden layers
        self.lin = nn.ModuleList()
        self.reparametrization = nn.ModuleList()
        for in_neurons, out_neurons in zip([input_dim] + self.num_neurons[:self.num_layers - 1], self.num_neurons):

            # Linear part associated to the offset
            if in_neurons == out_neurons or self.version == 'standard':
                # ... then no need to initialize
                self.reparametrization.append(nn.Identity())
            else:
                # ... then initialize with random weights on the simplex
                self.reparametrization.append(weight.linear_with_init(in_neurons, out_neurons, init='simplex'))

            # Classical linear part
            self.lin.append(weight.linear_with_init(in_neurons, out_neurons, init=init_param['hidden'],
                                                    weight_norm=weight_norm))
        self.activation, self.offset, self.range = act.get(activation_name=activation, version=version)

        # Output layer
        self.output = weight.linear_with_init(self.num_neurons[-1], output_dim, init=init_param['output'])

    def forward(self, xb):

        # Hidden layers
        for idl in range(self.num_layers):

            # Constraint on offset weights
            if self.version == 'bregman':
                self.reparametrization[idl] = weight.constraint(self.reparametrization[idl])

            # Perform forward pass
            if self.version == 'bregman':
                x_offset = torch.clamp(self.reparametrization[idl](xb), self.range[0], self.range[1])
                xb = self.activation(self.offset(x_offset) + self.lin[idl](xb))
            else:
                xb = self.activation(self.lin[idl](xb))

        # Output layer
        xb = self.output(xb)

        return xb
